### DISCUSSION TITLE
Improve update

### DISCUSSION THREAD

**Discussion by chochihim** [aarondl/sqlboiler#971] [Ideas]
Can `Update` be improved?

1. Allow single record be updated by columns (unique key maybe) other than primary key
2. Modify `UpdateAll` to use type-safe struct instead of `models.M` 
3. Utilise `RETURNING` clause so that `Update`, `UpdateAll` etc methods can return the struct/slice to caller

You can refer to https://github.com/volatiletech/sqlboiler/pull/970 for why I want to make this improvement

**Comment by stephenafamo** [comment]
I have a suggestion about how this can work.

**`UpdateAll` is modified to use a type-safe argument instead of `models.M{}`**

This directly solves point 2, but also solves point 1, because a single record can be updated using an appropriate query:
```go
_, err := models.UserWallets(
    models.UserWalletWhere.UserID.EQ(userID), // use whatever column matches
).UpdateAll(ctx, db, models.M{
    models.UserWalletColumns.Balance: balance,
})
```

## Returning

1. Add an extra argument to `UpdateAll` for binding.  
If this extra argument is not nil, a `RETURNING` clause is added and the result is bound to the argument.  
2. If only one record is expected, we can get `wallets[0]` as our single record. Alternatively, we can also add an `UpdateOne()` method to `UserWalletQuery` that adds `qm.Limit(1)`, auto-binds to a `UserWallet` struct and returns the value.

What do you think @aarondl ?

**Comment by aarondl** [comment]
@chochihim not against this. But what does the struct look like? There's a real difficulty in doing updates in both REST APIs and sql interfaces when it comes to "is this null? and I should update it? Or is this null and I should not?"

It requires 3 states:
* Omitted: Don't update this
* Null: Set this to null
* Value: Set this to value

Fortunately I could merge in a change that's in the null package that would make this possible and then we -could- use a struct.
See: https://github.com/volatiletech/null/pull/11
